pkg/controller/sync/status/status.go

```diff
@@ -104,51 +105,77 @@ type GenericFederatedStatus struct {
 
 type PropagationStatusMap map[string]PropagationStatus
 
+type CollectedPropagationStatus struct {
+	StatusMap        PropagationStatusMap
+	ResourcesUpdated bool
+}
+
 // SetPropagationStatus sets the conditions and clusters fields of the
-// federated resource's object map from the provided reason and
-// cluster status map.
-func SetPropagationStatus(fedObject *unstructured.Unstructured, reason AggregateReason, statusMap PropagationStatusMap) error {
+// federated resource's object map from the provided reason and collected
+// propagation status. Returns a boolean indication of whether status
+// should be written to the API.
+func SetPropagationStatus(fedObject *unstructured.Unstructured, reason AggregateReason, collectedStatus CollectedPropagationStatus) (bool, error) {
 	status := &GenericFederatedStatus{}
 	err := util.UnstructuredToInterface(fedObject, status)
 	if err != nil {
-		return errors.Wrapf(err, "Failed to unmarshall to generic status")
+		return false, errors.Wrapf(err, "Failed to unmarshall to generic status")
 	}
 	if status.Status == nil {
 		status.Status = &GenericPropagationStatus{}
 	}
-	propStatus := status.Status
 
-	// Identify whether one or more clusters could not be reconciled
-	// successfully.
-	if reason == AggregateSuccess && statusMap != nil {
-		for _, value := range statusMap {
-			if value != ClusterPropagationOK {
-				reason = CheckClusters
-				break
-			}
-		}
+	changed := status.Status.update(reason, collectedStatus)
+	if !changed {
+		return false, nil
 	}
-	propStatus.setPropagationCondition(reason)
-	propStatus.setClusterStatus(statusMap)
 
 	statusJSON, err := json.Marshal(status)
 	if err != nil {
-		return errors.Wrapf(err, "Failed to marshall generic status to json")
+		return false, errors.Wrapf(err, "Failed to marshall generic status to json")
 	}
 	statusObj := &unstructured.Unstructured{}
 	err = statusObj.UnmarshalJSON(statusJSON)
 	if err != nil {
-		return errors.Wrapf(err, "Failed to marshall generic status json to unstructured")
+		return false, errors.Wrapf(err, "Failed to marshall generic status json to unstructured")
 	}
 	fedObject.Object[util.StatusField] = statusObj.Object[util.StatusField]
 
-	return nil
+	return true, nil
+}
+
+// update ensures that the status reflects the given reason and collected
+// status. Returns a boolean indication of whether the status has been
+// changed.
+func (s *GenericPropagationStatus) update(reason AggregateReason, collectedStatus CollectedPropagationStatus) bool {
+	// Identify whether one or more clusters could not be reconciled
+	// successfully.
+	if reason == AggregateSuccess {
+		for _, value := range collectedStatus.StatusMap {
+			if value != ClusterPropagationOK {
+				reason = CheckClusters
+				break
+			}
+		}
+	}
+
+	clustersChanged := clusterStatusChanged(collectedStatus.StatusMap, s.Clusters)
```

Comment:
Can we somehow indicate that `clustersChanged` is in reference to their propagation status? Maybe consider renaming it to something like `clusterStatusChanged` or similar.

Reply:
The variable needs to have a different name than the function. Given that the usage of this variable is limited to 10 lines, what would you suggest?

Reply:
Oops, that was a typo on my part as I meant to add an `s` as in `clustersStatusChanged` but that's too similar perhaps. How about `clustersPropStatusChanged` or `clustersStatusHasChanged`?

Reply:
Done